src/pass.h

```diff
@@ -47,6 +51,7 @@ struct PassRegistry {
   std::vector<std::string> getRegisteredNames();
   std::string getPassDescription(std::string name);
   bool isPassHidden(std::string name);
+  bool doesPassAllowMultipleInstancesWithArgs(std::string name);
```

Comment:
Can we make this the behavior of all passes? I can't think of a problem that would occur, and that seems simpler, but again I might be missing something.

Reply:
Sure. I introduced the flag only for strict backwards compatibility for the other passes (bail out if the pass is specified multiple times with different arguments), but I'd be more than happy to remove it 😏

Reply:
Ok good, I don't think we need to bother with the case of passes passed different arguments before now (that was never meant to work). So if no tests break, then let's make that change.

Reply:
OK. Before I go ahead and make those changes, just to make sure we have the same picture of how the result should look like: All passes will use their own private value of the option that shares their name, and only options that have a different name and thus are set via `--pass-arg` will be read from the global map.

Reply:
Yes, that sounds right.

Reply:
Alright, I'll make the changes.

Reply:
Done. There were a few tests that were affected by the change to the semantics of `--pass-arg`  (cannot be used to set the "primary" argument anymore). If this is too invasive we can add the fallback to the global argument back again 😏

Reply:
Is this comment no longer valid?

https://github.com/WebAssembly/binaryen/blob/a12f18b1d6bfdcb565f755822e44594666a6cd0e/src/tools/optimization-options.h#L310-L317

It is not updated in this PR. If that behavior changes, please update it. However, I think we can keep that old behavior? I don't see a reason for a breaking change here (aside from corner cases of running a pass more than once with different arguments, which this PR should be improving).

Reply:
That's what I referred to with

> [...] and only options that have a different name and thus are set via --pass-arg will be read from the global map.

above 😏  Avoiding this breaking change was the reason why I added the fallback to the global option for in my first version of the PR. If that's fine with you I'll just add the fallback to the global version of the argument back to `getArgument` & friends, this will restore the old behaviour.

Reply:
Ah, I think I wasn't clear enough before, or I've misunderstood you. To make sure we are on the same page, the issue is that we need the fallback from the pass's argument to the global argument, because of the two forms `--foopass=FOO` and `--pass-arg=foopass@FOO`?

I think we can get around that. It looks like optimization-options handles the first form, and tool-options the second, but if they were coordinated this could be fixed, and passes would only read from the one place.

Concretely, I'm suggesting this:

* When we see `--foopass=FOO` then we check add it to the pass instance as an argument.
* When we see `--pass-arg=foopass@FOO` then we check if there is a pass for it.
  * If there is, we write it to that pass instance as an argument. If there is more than one appearance of the pass, set it for them all.
  * If there is no such pass, write it to the global store.

There is no need for a fallback here, since we always write to the pass instance when it has the name of a pass, and always to the global store when not.

That means some more processing in tool-options/optimization-options, but I think it's worth it. It would be simpler, and also backwards compatible except for corner cases with multiple invocations of the same pass (which I am ok with altering).

Reply:
Yeah, I think we managed to misunderstand each other 😏 

> To make sure we are on the same page, the issue is that we need the fallback from the pass's argument to the global argument, because of the two forms --foopass=FOO and --pass-arg=foopass@FOO?

Exactly.

I think your suggestion would work well, although I'd suggest a minor change:

* When we see `--foopass=FOO` we set the argument on the pass instance
* When we see `--pass-arg=foopass@FOO` we check the pass list for `foopass`
   - If we find an instance of `foopass` we set the argument there and do not continue to look for other, earlier instances
   - If there is no instance we store the argument in the global list, take it from there when we add `foopass` after     that and remove it from the global list afterwards

This way, `--pass-arg=foopass` will always affect the previous instance of `foopass` on the command line, except when it occurs before the first `--foopass`, in which case it will affect the first instance. To me, this is a little more consistent with how I would expect the UI to behave (instead of a single `--pass-arg=foopass` overriding all previous instances).

Reply:
Interesting. I like that it affects just one instance. Maybe we can make that even simpler, though: `--pass-arg=foopass@FOO` can always "attach" to the last instance, as you suggested, but also (1) it is an error if that pass already has an arg, and (2) it is an error if there is no pass instance before. Those avoid some possible ambiguity I think?

Reply:
I like that even better 😏 I'll make the necessary changes.

Reply:
OK, done. I also rebased my code on the current main head. If we agree that this is the behaviour we want to stick with I can also add a few tests.